cpp/src/arrow/type.h

```diff
@@ -126,7 +126,8 @@ struct ARROW_EXPORT DataTypeLayout {
 ///
 /// Simple datatypes may be entirely described by their Type::type id, but
 /// complex datatypes are usually parametric.
-class ARROW_EXPORT DataType : public detail::Fingerprintable {
+class ARROW_EXPORT DataType : public std::enable_shared_from_this<DataType>,
```

Comment:
Also not wonderful. I would rather add `DataType::Copy`, what do others think?

Reply:
I renamed GetSharedPtr to Copy and leaving the implementation which we can revisit later

Reply:
Using `enable_shared_from_this` sounds fine to me. There is basically no use case for creating non-shared-ptr'ed DataTypes anyway.